csharp/src/Google.Protobuf/ParsingPrimitives.cs

```diff
@@ -638,17 +638,64 @@ public static string ReadRawString(ref ReadOnlySpan<byte> buffer, ref ParserInte
             }
 #endif
 
-            var decoder = WritingPrimitives.Utf8Encoding.GetDecoder();
+            return ReadStringSlow(ref buffer, ref state, length);
+        }
 
-            // TODO: even if GOOGLE_PROTOBUF_SUPPORT_FAST_STRING is not supported,
-            // we could still create a string efficiently by using Utf8Encoding.GetString(byte[] bytes, int index, int count)
-            // whenever the buffer is backed by a byte array (and avoid creating a new byte array), but the problem is
-            // there is no way to get the underlying byte array from a span.
+        /// <summary>
+        /// Reads a string assuming that it is spread across multiple spans in a <see cref="ReadOnlySequence{T}"/>.
+        /// </summary>
+        private static string ReadStringSlow(ref ReadOnlySpan<byte> buffer, ref ParserInternalState state, int length)
+        {
+#if GOOGLE_PROTOBUF_SUPPORT_FAST_STRING
+            if (IsDataAvailable(ref state, length))
+            {
+                // Read string data into a temporary buffer, either stackalloc'ed or from ArrayPool
+                // Once all data is read then call Encoding.GetString on buffer and return to pool if needed.
+
+                byte[] byteArray = null;
+                Span<byte> byteSpan = length <= StackallocThreshold ?
+                    stackalloc byte[length] :
+                    (byteArray = ArrayPool<byte>.Shared.Rent(length));
 
-            // TODO: in case the string spans multiple buffer segments, creating a char[] and decoding into it and then
-            // creating a string from that array might be more efficient than creating a string from the copied bytes.
+                try
+                {
+                    int remainingByteLength = length;
```

Comment:
the code that reads possibly multiple segments (refreshing when needed) into a span is doing exactly the same thing as one of the branches of ReadRawBytes (after we've checked that the data is available and that is safe to allocate the span) - I think we should refactor that code out into a private method.
Doing that should also help fix some inconsistencies
- e.g. currently for ReadStringSlow/ReadRawString we don't seem to perform a check for crossing currentLimit (while ReadRawBytes performs such check) - it would be good to unify.
- also the exact "IsDataAvailable" check details are slightly different in ReadRawBytes (it checks `size < buffer.Length || size < state.segmentedBufferHelper.TotalLength`  which works but IsDataAvailable arguably does a better job at that - at least in terms of the `size < state.segmentedBufferHelper.TotalLength` (as opposed to `size < state.segmentedBufferHelper.TotalLength - state.totalBytesRetired - state.bufferPos` from IsDataAvailable) part of the condition.